File: tps/tp01_2026/IsRecursivo/IsRecursivo.c
#include <stdio.h>
#include <string.h>

int ehVogal(char c) {
    return c=='a'||c=='e'||c=='i'||c=='o'||c=='u'||
           c=='A'||c=='E'||c=='I'||c=='O'||c=='U';
}

int ehLetraASCII(char c) {
    return (c>='a'&&c<='z')||(c>='A'&&c<='Z');
}

int somenteVogais(char *s, int i) {
    if (s[i] == '\0') return 1;
    if (!ehLetraASCII(s[i]) || !ehVogal(s[i])) return 0;
    return somenteVogais(s, i + 1);
}

int somenteConsoantes(char *s, int i) {
    if (s[i] == '\0') return 1;
    if (!ehLetraASCII(s[i]) || ehVogal(s[i])) return 0;
    return somenteConsoantes(s, i + 1);
}

int ehInt(char *s, int i, int qtd) {
    if (s[i] == '\0') return qtd > 0;
    if (s[i] < '0' || s[i] > '9') return 0;
    return ehInt(s, i + 1, qtd + 1);
}

int ehReal(char *s, int i, int temPonto, int temDigito) {
    if (s[i] == '\0') return temPonto && temDigito;
    if (s[i] >= '0' && s[i] <= '9') return ehReal(s, i+1, temPonto, 1);
    if (s[i] == '.' && !temPonto)    return ehReal(s, i+1, 1, temDigito);
    return 0;
}

int main() {
    char linha[10000];
    while (fgets(linha, sizeof(linha), stdin)) {
        int len = strlen(linha);
        if (len > 0 && linha[len-1] == '\n') linha[--len] = '\0';
        if (strcmp(linha, "FIM") == 0) break;

        int sv = len > 0 && somenteVogais(linha, 0);
        int sc = len > 0 && somenteConsoantes(linha, 0);
        int ei = ehInt(linha, 0, 0);
        int er = ehReal(linha, 0, 0, 0);

        printf("%s %s %s %s\n",
            sv ? "SIM" : "NAO",
            sc ? "SIM" : "NAO",
            ei ? "SIM" : "NAO",
            er ? "SIM" : "NAO");
    }
    return 0;
}

File: tps/tp01_2026/IsRecursivo/IsRecursivo.java
public class IsRecursivo {

    static boolean ehFIM(String s) {
        return s.length() == 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M';
    }

    static boolean ehVogal(char c) {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
    }

    static boolean ehLetraASCII(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
    }

    static char toLower(char c) {
        return (c >= 'A' && c <= 'Z') ? (char)(c + 32) : c;
    }

    static boolean somenteVogais(String s, int i) {
        if (i == s.length()) return i > 0;
        char c = toLower(s.charAt(i));
        if (!ehLetraASCII(c) || !ehVogal(c)) return false;
        return somenteVogais(s, i + 1);
    }

    static boolean somenteConsoantes(String s, int i) {
        if (i == s.length()) return i > 0;
        char c = toLower(s.charAt(i));
        if (!ehLetraASCII(c) || ehVogal(c)) return false;
        return somenteConsoantes(s, i + 1);
    }

    static boolean ehInt(String s, int i) {
        if (i == s.length()) return i > 0;
        char c = s.charAt(i);
        if (c < '0' || c > '9') return false;
        return ehInt(s, i + 1);
    }

    static boolean ehReal(String s, int i, boolean temPonto, boolean temDigito) {
        if (i == s.length()) return temPonto && temDigito;
        char c = s.charAt(i);
        if (c >= '0' && c <= '9') return ehReal(s, i + 1, temPonto, true);
        if (c == '.' && !temPonto) return ehReal(s, i + 1, true, temDigito);
        return false;
    }

    public static void main(String[] args) {
        String linha = MyIO.readLine();
        while (!ehFIM(linha)) {
            MyIO.println(
                (somenteVogais(linha, 0) ? "SIM" : "NAO") + " " +
                (somenteConsoantes(linha, 0) ? "SIM" : "NAO") + " " +
                (ehInt(linha, 0) ? "SIM" : "NAO") + " " +
                (ehReal(linha, 0, false, false) ? "SIM" : "NAO")
            );
            linha = MyIO.readLine();
        }
    }
}
